tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11436,6 +11492,7 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
         Diags.Report(pPatchFnDecl->getSourceRange().getBegin(), id);
       }
     }
+    ValidateEntryPointFunctionAttributes(self, pEntryPointDecl);
```

Comment:
The reason I can't move it under Diagnose Entry is that it doesn't always correctly identify the entry point decl. For example, in WaveSizeAttrNotAllowed2.hlsl, it will treat main2 as the entry point when really main is the entry point. This causes a diagnostic misfire.

Reply:
The reason is because there are non-entry point function declarations that will get DiagnoseEntry called on them. The function seems poorly named, since it will be called on non-entry point functions.

Reply:
`DiagnoseEntry` is called on functions that *could* be an entry, then it determines whether it *is* an entry (by looking up shader attribute, after adding it if necessary), at which point, we can diagnose entry attributes, since we know it's an entry function.